text/0000-app-mixins.md

```diff
@@ -0,0 +1,243 @@
+# Meta
+[meta]: #meta
+- Name: Application Mixins
+- Start Date: 2020-06-22
+- Author(s): [Joe Kutner](https://github.com/jkutner/)
+- RFC Pull Request: (leave blank)
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: N/A
+
+# Summary
+[summary]: #summary
+
+This is a proposal for allowing application developers (buildpack users) to specify mixins that will be dynamically included with their build and run images.
+
+# Motivation
+[motivation]: #motivation
+
+Mixins already allow buildpack authors to create buildpacks that depend on an extended set of OS packages without affecting build time, but it's common for application code to depend on OS packages too. But allowing buildpacks to arbitrarily install OS packages during build would drastically increase build time, especially when CNB tooling is used to build or rebase many apps with similar package requirements.
+
+For that reason, this proposal defines a mechanism to implement dynamic installation of mixins at build time while minimizing the performance impact. This is accomplished by allowing stack images to satisfy mixin requires statically and/or reject dynamic mixin installation early in the build process.
+
+# What it is
+[what-it-is]: #what-it-is
+
+- *application developer* - a person who is using buildpacks to transform their application code into an OCI image
+- *root buildpack* - a new type of buildpack that runs as the root user
+- *stackpack* - a type of root buildpack that runs against the stack image(s) instead of an app. It is distinguished by a static list of mixins it can provide.
+
+An application developer may specify a list of mixins in their application's `project.toml` file like this:
+
+```toml
+[build]
+mixins = [ "<mixin name>" ]
+```
+
+When a command like `pack build` is run, the list of mixins will be processed before buildpacks are run. For each mixin name, the following will happen:
+
+* If the mixin name is prefixed with `build:` (as per [RFC-0006](https://github.com/buildpacks/rfcs/blob/main/text/0006-stage-specific-mixins.md)) the mixin will be dynamically added to the build image.
```

Comment:
Newbie question: I'm assuming this happens prior to any phase of the buildpack lifecycle happening. As in, it happens when we `pack create-builder ...`. Which means all the buildpack lifecycle steps will never read to read/write from/to `/`. Is this accurate?

Reply:
(By lifecycle steps I mean: detector, analyzer, restorer, builder, exporter)

Reply:
No, the stackpacks would not execute during `create-builder`. They run during the normal lifecycle phases, but in as the `root` user, which allows them to write to `/`). Does that help?

Reply:
I see. So in the line above you mean the mixin will be dynamically added to the builder* image when prefixed with `build:`. 

I'm curious, if implemented, which lifecycle phase will need to read/write to/from `/`. All of them or just the `builder` phase?

Reply:
Only the build phase will need (or more correctly "be able") to write to `/`